src/io/iohk/decco/Codec.scala

```diff
@@ -2,7 +2,28 @@ package io.iohk.decco
 
 import java.nio.ByteBuffer
 
-trait Codec[T] { self =>
+trait Codec[T] {
+
+  import Codec.Failure
+
+  def encode[B](t: T)(implicit bi: BufferInstantiator[B]): B
+
+  def decode[B](start: Int, source: B)(implicit bi: BufferInstantiator[B]): Either[Failure, T]
+
+  def decode[B](source: B)(implicit bi: BufferInstantiator[B]): Either[Failure, T]
+}
+
+object Codec {
+
+  case class DecodeResult[+T](decoded: T, nextIndex: Int)
+  case object Failure
+  type Failure = Failure.type
+
+  def apply[T](implicit ev: CodecContract[T]): CodecContract[T] = ev
+
+}
+
+trait CodecContract[T] { self =>
```

Comment:
I can see that `CodecContract` implements `Codec` interface and there is a definition for an implicit conversion from the former to the later (defined in auto package object) that just calls the implemented methods.
Is there any reason not to simply extend `Codec` here and avoid implicit conversion completely? or, is there a case where you would like `CodecContract` not to be available where `Codec` is required?

Reply:
@EzequielPostan That would make function taking `Codec` as implicit parameter see implicit `CodecContract`s as viable values, leading to ambiguity in the case where we want Snappy to wrap such `CodecContract` or even pure `CodecContract` silently winning over transformed `Codec` (this occured in my tests).

Maybe it can be fixed somehow, but I'm not familiar enough with Scala's implicits resolution order to find a solution and definitely not no be confident that it won't silently break in some contexts.

Reply:
Thank you for the explanation. 
It sounds reasonable to me, I don't have an alternative solution to offer.